examples/benchmark.py

```diff
@@ -0,0 +1,311 @@
+"""
+
+import argparse
+import collections
+import logging
+import time
+
+import hmmlearn.hmm
+
+import numpy as np
+
+import sklearn.base
+
+LOG = logging.getLogger(__file__)
+
+
+class Benchmark:
+    def __init__(self, repeat, n_iter, verbose):
+        self.repeat = repeat
+        self.n_iter = n_iter
+        self.verbose = verbose
+
+    def benchmark(self, sequences, lengths, model, tag):
+        elapsed = []
+        for i in range(self.repeat):
+            start = time.time()
+            cloned = sklearn.base.clone(model)
+            cloned.fit(sequences, lengths)
+            end = time.time()
+            elapsed.append(end-start)
+            self.log_one_run(start, end, cloned, tag)
+        return np.asarray(elapsed)
+
+    def generate_training_sequences(self):
+        pass
+
+    def new_model(self, implementation):
+        pass
+
+    def run(self, results_file):
+        runtimes = collections.defaultdict(dict)
+
+        sequences, lengths = self.generate_training_sequences()
+
+        for implementation in ["scaling", "log"]:
+            model = self.new_model(implementation)
+            LOG.info(f"{model.__class__.__name__}: testing {implementation}")
+            key = f"{model.__class__.__name__}|EM|hmmlearn-{implementation}"
+            elapsed = self.benchmark(sequences, lengths, model, key)
+            runtimes[key]["mean"] = elapsed.mean()
+            runtimes[key]["std"] = elapsed.std()
+
+        with open(results_file, mode="w") as fd:
+            fd.write("configuration,mean,std,n_iterations,repeat\n")
+            for key, value in runtimes.items():
+                fd.write(f"{key},{value['mean']},{value['std']},{self.n_iter},{self.repeat}\n")
+
+    def log_one_run(self, start, end, model, tag):
+        LOG.info(f"Training Took {end-start} seconds {tag}")
+        LOG.info(f"startprob={model.startprob_}")
+        LOG.info(f"transmat={model.transmat_}")
+
+
+class GaussianBenchmark(Benchmark):
+
+    def new_model(self, implementation):
+        return hmmlearn.hmm.GaussianHMM(
+            n_components=4,
+            n_iter=self.n_iter,
+            covariance_type="full",
+            implementation=implementation,
+            verbose=self.verbose
+        )
+
+    def generate_training_sequences(self):
+        pi = np.asarray([0, 0, 0, 1])
+        A = np.asarray([
+            [.2, .2, .3, .3],
+            [.3, .2, .2, .3],
+            [.2, .3, .3, .2],
+            [.3, .3, .2, .2],
+        ])
+        means = np.asarray([
+            -1.5,
+            0,
+            1.5,
+            3
+        ]).reshape(4, 1)
+        variances = np.asarray([
+            .5,
+            .5,
+            .5,
+            .5
+        ]).reshape(4, 1, 1,)
+
+        sampler = hmmlearn.hmm.GaussianHMM(
+            n_components=4,
+            covariance_type="full",
+            init_params="",
+            verbose=self.verbose
+        )
+        sampler.startprob_ = pi
```

Comment:
I would directly define the arrays there, instead of doing it in two steps (same comment applies throughout).

Reply:
This should all be done now.